### DISCUSSION TITLE
Support alphabetical / lexicographical versioning scheme

### DISCUSSION THREAD

**Discussion by mikaello** [renovatebot/renovate#23477] [Suggest an Idea]
Timestamp-based versioning, where every commit is released as a new timestamp (+ maybe the hash of the commit) is not an uncommon versioning. To determine if a version is newer than another, lexical order is used, this is currently not supported by Renovate.

Example of a docker image with time-based versioning (this is just one example, the exact syntax of the timestamp may vary): `ghcr.io/nais/azurerator:2023-07-14-095208-25ad30c`

To get around this in Renovate the regex versioning could be used to make the versioning fit into the SemVer box, but this has several problems:
* There is no SemVer semantics in timestamp-based versioning, giving a false impression of semantics when a major, minor, patch, or build release is incoming
* Not supporting more than 5 elements (major/minor/patch/build/revision), so in a "normal" timestamp this will mean Renovate can't differentiate releases released within the same minute (given `2023-20-07T16-30-20.fe8a923`, best effort SemVer will be: 2023 (year) = major, 20 (month) = minor, 7 (day) = patch, 16 (hour) = build, 30 (minute) = revision)
* Regexes are error-prone and hard to review
* The versioning is hard to reuse, possibly leading to copy-paste and drifting if you fix something in one place, but forget other places
* if your dependencies use slightly different formats for their timestamp-based versioning you may need to create multiple regex versions to match all of them

<details><summary>Click to see an example of how regex SemVer workaround could look like</summary>

Renovate config (just an example with a regex matcher, see the "YAML file" below for the regex versioning:

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "regexManagers": [
    {
      "fileMatch": [".*"],
      "matchStrings": [
        "renovate:\\s*datasource=(?<datasource>[a-z-]+?) depName=(?<depName>\\S+?)(?: (packageName)=(?<packageName>\\S+?))?(?: versioning=(?<versioning>\\S+?))?(?: registryUrl=(?<registryUrl>\\S+?))?\\n\\s*(?:\\S+?_)?(?:VERSION|version|TAG|tag)\\s*[?]?[=:]\\s*\"?(?<currentValue>\\S+?)\"?\\s"
      ],
      "versioningTemplate": "{{#if versioning}}{{{versioning}}}{{else}}semver{{/if}}"
    },
  ]
}
```

Some YAML file with a dependency defined:

```yaml
# renovate: datasource=docker depName=kafka/cruise-control versioning=regex:^(?<major>\d+)-(?<minor>\d+)-(?<patch>\d+)T(?<build>\d+)-(?<revision>\d+)$
regex_cruise_control_version: 2022-12-27T04-08-07.a2fe6049
```

</details>

This would be solved by providing a versioning scheme based on lexicographical order. It already exists some workarounds in the code that could be modified to support, e.g. in [AWS Machine Image](https://github.com/renovatebot/renovate/blob/f19b3001847be370f50d1ed5253e726b55391477/lib/modules/versioning/aws-machine-image/index.ts#L14-L29) or [Git versioning](https://github.com/renovatebot/renovate/blob/f19b3001847be370f50d1ed5253e726b55391477/lib/modules/versioning/git/index.ts#L12-L23), if we had just compared `suffix` in `_compare()`, wouldn't that be sufficient?

There is also lexicographical sort order implemented in the [deb versioning](https://github.com/renovatebot/renovate/blob/f19b3001847be370f50d1ed5253e726b55391477/lib/modules/versioning/deb/index.ts#L111), so there is some prior effort to do this.

How to compare timestamp-based versioning is a question that has been asked multiple times before (e.g. https://github.com/renovatebot/renovate/discussions/13020 and https://github.com/renovatebot/renovate/discussions/17879), so I think a better solution than the regex versioning is worth looking into :-)

**Comment by rarkins** [comment]
There is already "loose" versioning which should sort correctly.

**Reply by mikaello** [reply]
You are right, thanks!

I have been reading through the [versioning/loose source code](https://github.com/renovatebot/renovate/blob/a89e307af40eddbc9fe626bfb61d65b667e0a530/lib/modules/versioning/loose/index.ts) before, but wrongly concluded that it didn't work, and which was reinforced by replies in similar issues.

Would you be open to adding a few tests in 'loose' with some timestamp cases? I.e. https://github.com/renovatebot/renovate/commit/11e9ef38e7ed1c1c396f853ffe0a94a90cda478b . I can create a PR if OK.

UPDATE - PR opened: https://github.com/renovatebot/renovate/pull/23507

**Reply by rarkins** [reply]
Further tests would be welcome, thanks
